Win_firewall_profile Module

lowlighter opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Currently, modules `win_firewall` and `win_firewall_rule` allows to manage Windows firewall through Ansible. However, it is not yet possible to set a default action for inbound/outbound connections (special rules which are triggered when no other rules for a given profile matched). 

As Windows firewall rules aren't really ordered, rules likes *block all unwanted connections* cannot be applied by creating rules with `win_firewall_rule`, hence this submission.

Also, logging of allowed/blocked/ignored packets can be useful so this module also add these options.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
win_firewall_profile

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Current behavior :
```yaml
- name: Block unwanted inbound connections for Domain and Private profiles
  win_shell: Set-NetFirewallProfile -name Domain,Private -DefaultInBoundAction Block
```
Proposed behavior : 
<!--- Paste verbatim command output below, e.g. before and after your change -->
```yaml
- name: Block unwanted inbound connections for Domain and Private profiles and log blocked packets to custom file
  win_firewall_profile:
    inbound: Block
    log_blocked: True
    log_file: "C:\\myfirewall.log"
    profiles:
      - Domain
      - Private
```

lowlighter committed:
Add Windows firewall profile manager module

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.ps1:
```diff
@@ -0,0 +1,60 @@
+
+$ErrorActionPreference = "Stop"
+$firewall_profiles = @("Domain", "Private", "Public")
+$firewall_actions = @("NotConfigured", "", "Allow", "", "Block")
```
Why not make it an enum or a hashtable?

lowlighter replied:
Used toString() method instead as it already translates options to the correct string value

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.ps1:
```diff
@@ -0,0 +1,60 @@
+
+    }
+} Catch {
+    Fail-Json $result "an error occurred when attempting to change firewall profile $profile $($_.Exception.Message)"
```
If I'm not mistaken on recent ansible versions this general exception catching is not needed and might even make debugging harder for people

lowlighter replied:
Removed general catch

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.ps1:
```diff
@@ -0,0 +1,60 @@
+
+    ForEach ($profile in $firewall_profiles) {
+
+        $currentinbound = (Get-NetFirewallProfile -Name $profile).DefaultInBoundAction
```
Personally I prefer saving the output and not keep calling the same Get

lowlighter replied:
Now saving the output so get is called once

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.ps1:
```diff
@@ -0,0 +1,60 @@
+            outbound = $firewall_actions[$currentoutbound]
+        }
+
+        if ($profiles -notcontains $profile) {
```
IMHO continue makes things look weird in the code, think about just using the above check you make for considered with the two later ifs in it

lowlighter replied:
Changed the way it worked so it only iterates through given profiles instead of all profiles

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.py:
```diff
@@ -0,0 +1,56 @@
+---
+module: win_firewall_profile
+version_added: "2.10"
+short_description: Windows firewall profile automation
```
Maybe 'Windows firewall profile default action automation'

lowlighter replied:
Updated description

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.py:
```diff
@@ -0,0 +1,56 @@
+version_added: "2.10"
+short_description: Windows firewall profile automation
+description:
+  - Allows you to update settings for Windows firewall profiles.
```
Maybe mention it's the default action again

lowlighter replied:
Same here

ShachafGoldstein reviewed (commented):
Please add a changelog fragment and tests

ShachafGoldstein commented on lib/ansible/modules/windows/win_firewall_profile.py:
```diff
@@ -0,0 +1,56 @@
+    outbound: NotConfigured
+'''
+
+RETURN = r'''
```
Please fill in the return part as well

lowlighter replied:
Filled

lowlighter commented:
Thanks for taking your time to review this PR.

I have refactored the code to be compliant with what you said and followed more closely the [official documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general_windows.html#windows-new-module-development) for developing Windows module.

I added additional options to configure logging of packets for each firewall profile and created the  integration tests. However, I didn't make a changelog fragment as it seems it is done [automatically](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs) for new modules.

Have a nice day.

lowlighter commented:
ready_for_review
